### DISCUSSION TITLE
How to update for 4.1

### DISCUSSION THREAD

**Discussion by Pesegato** [dyn4j/dyn4j#184] [Q&A] [answered]
I have this code that I am updating for 4.1, and I am not sure on the replacement for `getFixture1()`

```
    BroadphaseDetector<Body> bp;
            List<CollisionPair<Body>> pairs = bp.detect();
            for (CollisionPair<Body> pair : pairs) {
                Body body1 = pair.getFirst();
                Body body2 = pair.getSecond();
                BodyFixture fixture1 = pair.getFixture1();
                BodyFixture fixture2 = pair.getFixture2();
                Transform transform1 = body1.getTransform();
                Transform transform2 = body2.getTransform();
                Convex convex2 = fixture2.getShape();
                Convex convex1 = fixture1.getShape();
            Penetration p = new Penetration();
            if (np.detect(convex1, transform1, convex2, transform2, p)) {
                    //System.out.println("Collision " + fixture1.getUserData() + " " + fixture2.getUserData());
                    for (CollisionListener listener:listeners){
                        listener.listen((Long)fixture1.getUserData(), (Long)fixture2.getUserData());
                    }
                }
            }
```

For reference, here is my previous request for 4.0
https://github.com/dyn4j/dyn4j/issues/144

Thanks!

**Comment by NordelDev** [comment]
Hi,

as far as I know you have to use `CollisionPair<CollisionItem<SimulationBody, BodyFixture>>>`. Then you can use `collisionPair.getFirst().getBody()` or `...getFixture()`.

See https://github.com/dyn4j/dyn4j/discussions/183 for reference.

**Comment by wnbittle** [comment, marked as answer]
Building a broadphase detector is more involved now.  You must supply a filter, aabb producer, and expansion method.  Default implementations of these exist and are used in the `CollisionWorld`, `PhysicsWorld`, and `World` classes.  Using the `CollisionWorld` abstract class would abstract away a lot of this setup.

```java
final BroadphaseFilter<CollisionItem<Body, BodyFixture>> broadphaseFilter = new CollisionItemBroadphaseFilter<Body, BodyFixture>();
final AABBProducer<CollisionItem<Body, BodyFixture>> aabbProducer = new CollisionItemAABBProducer<Body, BodyFixture>();
final AABBExpansionMethod<CollisionItem<Body, BodyFixture>> expansionMethod = new StaticValueAABBExpansionMethod<CollisionItem<Body, BodyFixture>>(0.2);
final BroadphaseDetector<CollisionItem<Body, BodyFixture>> broadphase = new DynamicAABBTree<CollisionItem<Body, BodyFixture>>(
	broadphaseFilter,
	aabbProducer, 
	expansionMethod, 
	initialBodyCapacity);
final CollisionItemBroadphaseDetector<Body, BodyFixture> broadphaseDetector = new CollisionItemBroadphaseDetectorAdapter<Body, BodyFixture>(broadphase);
broadphaseDetector.setUpdateTrackingEnabled(true);

List<CollisionPair<CollisionItem<Body, BodyFixture>>> pairs = broadphaseDetector.detect();
for (CollisionPair<CollisionItem<Body, BodyFixture>> pair : pairs) {
    // handle the pairs by using pair.getFirst().getBody() / pair.getSecond().getFixture() / etc.
}
```

**Reply by Pesegato** [reply]
But I still need the NarrowPhase, right? The full code translates into:

```
            List<CollisionPair<CollisionItem<Body, BodyFixture>>> pairs = broadphaseDetector.detect();
            for (CollisionPair<CollisionItem<Body, BodyFixture>> pair : pairs) {
                // handle the pairs by using pair.getFirst().getBody() / pair.getSecond().getFixture() / etc.
                BodyFixture fixture1 = pair.getFirst().getFixture();
                BodyFixture fixture2 = pair.getSecond().getFixture();
                Transform transform1 = pair.getFirst().getBody().getTransform();
                Transform transform2 = pair.getSecond().getBody().getTransform();
                Convex convex2 = fixture2.getShape();
                Convex convex1 = fixture1.getShape();
            Penetration p = new Penetration();
            if (np.detect(convex1, transform1, convex2, transform2, p)) {
                    //System.out.println("Collision " + fixture1.getUserData() + " " + fixture2.getUserData());
                    for (CollisionListener listener:listeners){
                        listener.listen((Long)fixture1.getUserData(), (Long)fixture2.getUserData());
                    }
                }
            }
```

**Reply by Pesegato** [reply]
Also: 

2) how do I add Bodies to the Broadphase? This won't work anymore:

` broadphase.add(body);`

3) body.setAutoSleepingEnabled is replaced with body.setAtRestDetectionEnabled; is this correct?

4) How about this?

```
            for (Body b : bodies) {
                broadphase.update(b);
            }
```

Thanks!

**Reply by wnbittle** [reply]
> But I still need the NarrowPhase, right?

Yes.  Nothing changed with the output of the broadphase, just the way it's being used.

> 2. how do I add Bodies to the Broadphase? This won't work anymore:
`broadphase.add(body);`

If you are using the [`CollisionItemBroadphaseDetector`](https://github.com/dyn4j/dyn4j/blob/master/src/main/java/org/dyn4j/collision/broadphase/CollisionItemBroadphaseDetector.java) interface there's a [`add(Body)`](https://github.com/dyn4j/dyn4j/blob/master/src/main/java/org/dyn4j/collision/broadphase/CollisionItemBroadphaseDetector.java#L69) method.

> 3. body.setAutoSleepingEnabled is replaced with body.setAtRestDetectionEnabled; is this correct?

Yes

> 4. How about this?
```java
for (Body b : bodies) {
    broadphase.update(b);
}
```

Yes.  You can also do [`broadphase.update()`](https://github.com/dyn4j/dyn4j/blob/master/src/main/java/org/dyn4j/collision/broadphase/BroadphaseDetector.java#L111) now instead of iterating all the bodies again (assuming fixtures have not be added to the bodies).

A lot of this is handled in the [`AbstractCollisionWorld`](https://github.com/dyn4j/dyn4j/blob/master/src/main/java/org/dyn4j/world/AbstractCollisionWorld.java) abstract class.  You may want to consider extending that class instead of doing all this manually.
